lib/std/unicode.zig

```diff
@@ -631,3 +631,32 @@ test "utf8ToUtf16LeWithNull" {
         testing.expect(utf16[2] == 0);
     }
 }
+
+pub fn utf16leStringLiteral(comptime str: [*:0]const u8) [*c]const c_ushort {
```

Comment:
The resulting string should be null-terminated, no?
And `u16` looks much better than `c_ushort`

Reply:
Result type should ideally be a single-item pointer to a null-terminated array of u16 (e.g. same as normal string literals, but u16 instead of u8).

`*[N:null]const u16`

Reply:
The resulting string should be null-terminated. 

`u16` vs `c_ushort` was one of my concerns. I would prefer `u16`, which will make it interface cleanly with Windows functions declared using `LPCWSTR` (but not with imported C functions like `wprintf`). All existing Unicode functions use `u16`. Unsure of the pros/cons for this.

Reply:
u16 will interface cleanly with c_ushort. they're the same on all windows systems

Reply:
FWIW nt's `UNICODE_STRING` does not require null termination (and infact paths can contain null bytes!).

Reply:
Yep, but isn't this function targeting the other use case? I'm sure both would be useful.

Reply:
> Yep, but isn't this function targeting the other use case? I'm sure both would be useful.

Yes. There are many APIs on windows that require a null terminated string. I just wanted to note that the ntapi ones rarely do. As well as note that `mem.len` would be inappropriate for some APIs due to nt paths being allowed to contain a null byte.

Reply:
Perhaps I'm testing i16/c_ushort interactions incorrectly, then? Whenever I try to call `c.wprintf` with the modified function on Windows I get a compilation error:

> .\src\main.zig:19:21: error: expected type '[*c]const c_ushort', found '[*c]const u16'
    _ = c.wprintf(W("Hello World\n"));
                    ^
.\src\main.zig:19:21: note: pointer type child 'u16' cannot cast into pointer type child 'c_ushort'
    _ = c.wprintf(W("Hello World\n"));
                    ^
.\src\main.zig:19:21: note: unsigned 16-bit int cannot represent all possible unsigned 16-bit values
    _ = c.wprintf(W("Hello World\n"));

Reply:
huh. well "unsigned 16-bit int cannot represent all possible unsigned 16-bit values" is a false statement made by the compiler, and that's definitely a bug!

I'll take a look, thanks

Reply:
I don't quite understand how `*[N:null]const u16` would be used. I wasn't able to get `*[:0]const u16` to compile when supplied for as a LPCWSTR parameter.

Reply:
`*[:0]const u16` is a pointer to a slice whose pointer is null terminated. i.e. a double pointer.

`*[N:0]const u16` is a pointer to an array which has a sentinel 0 value. i.e. a single pointer

`[*:0]const u16` is a null terminated pointer. This is what `LPCWSTR` is.

`*[N:0]const u16` coerces to `[*:0]const u16`. 

`*[:0]const u16` is a red herring :red_circle:  :fish: